File: DSAjavascript/BinarySearch/SearchRotatedSortArray2.js
/**
 * leetcode problem no :-81
 * leetcode solution link(https://leetcode.com/problems/search-in-rotated-sorted-array-ii/)
 * 
 */
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {boolean}
 */
var search = function(nums, target) {

    let start = 0;
    let end = nums.length-1;

    while(start <= end){

        let mid = start + Math.floor((end-start)/2);

        if(nums[mid] == target){
            return true;
        }
        if(nums[mid] == nums[start]){

            start ++;
            continue;
        }
        //Here check the condition in left side target is present or not

        if(nums[mid] > nums[start]){
            if(target < nums[mid] && target >= nums[start]){

                end = mid -1;
            }else{
                start = mid +1;
            }
        }
        // Here check in the right side
        else{
            if(target > nums[mid] && target <= nums[end]){

                start = mid +1;
            }else{
                end = mid -1;
            }
        }
    }
    return false;
};

File: DSAinJava/search.java/SearchRotatedSortArray2.java
/**
 * leetcode problem no :- 81
 * leetcode solution link(https://leetcode.com/problems/search-in-rotated-sorted-array-ii/)
 */

 class Solution {
    public boolean search(int[] nums, int target) {

        int s =0;
        int e= nums.length -1;

        while(s <= e){

            int mid = s+ (e-s)/2;

            if(nums[mid] == target){
                return true;
            }
            if(nums[mid ] == nums[s]){

                s++;
                continue;
            }
            // check target is left side
            if(nums[s] < nums[mid]){
                if(target < nums[mid] && target >= nums[s]){
                    e =mid -1;
                }else{
                    s = mid +1;
                }
               
                } else{
                    //check target is right side 
                    if(target > nums[mid] && target <= nums[e]){

                        s = mid +1;
                    }else{

                        e =mid-1;
                        
                    }
            }
        }
        return false;
    }
}